platform-plugins.md

```diff
@@ -0,0 +1,305 @@
+    batteryLevel = "Failed to get battery level: '${e.message}'.";
+  }
+```
+
+**Step 2a: Add an iOS platform-specific implementation**
+
+*TODO*
+
+**Step 2b: Add an Android platform-specific implementation**
+
+Start by opening the Android host portion of your Flutter app in Android Studio:
+
+1. Start Android Studio
+
+1. Invoke File > Open...
+
+1. Navigate to the directory holding your Flutter app, and select the `android`
+folder inside it. Click OK.
+
+1. Open the file `MainActivity.java` located in the `java` folder in the Project
+view.
+
+Next, create a `FlutterMethodChannel` and set a `MethodCallHandler` inside the
+`onCreate` method. Make sure to use the same channel name as was used on the
+Flutter client side.
+
+```java
+@Override
+public void onCreate(Bundle savedInstanceState) {
+  private static final String CHANNEL = "battery";
+
+  super.onCreate(savedInstanceState);
+
+  new FlutterMethodChannel(getFlutterView(), CHANNEL).setMethodCallHandler(
+      new MethodCallHandler() {
+        @Override
+        public void onMethodCall(MethodCall call, Response response) {
+          // TODO
+        }
+  });
+}
+```
+
+Next, we add the actual Android Java code that uses the Android battery APIs to
+retrieve the battery level. This code is exactly the same you would have written
+in a native Android app. Add the following as a new method in the activity
+class, below the `onCreate` method:
+
+```java
+private int getBatteryLevel() {
+  int batteryLevel = -1;
+  if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP) {
+    BatteryManager batteryManager = (BatteryManager) getSystemService(BATTERY_SERVICE);
+    batteryLevel = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
+  } else {
+    Intent intent = new ContextWrapper(getApplicationContext()).
+        registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
+    batteryLevel = (intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) * 100) /
+        intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1);
+  }
+
+  return batteryLevel;
+}
+```
+
+Finally, we add handlers for all supported methods, in our case the single
+method `getBatteryLevel`, which simply calls the Android code we wrote in the
+previous step.
+
+```java
+       @Override
+       public void onMethodCall(MethodCall call, Response response) {
+         if (call.method.equals("getBatteryLevel")) {
+           int batteryLevel = getBatteryLevel();
+
+           if (batteryLevel != -1) {
+             response.success(batteryLevel);
+           } else {
+             response.error("UNAVAILABLE", "Battery level not available.", null);
+           }
+
+         } else {
+           throw new IllegalArgumentException("Unknown method " + call.method);
+         }
+       }
+```
+
+Note how we use the `response` argument to pass back both a response for both
+the success and error cases, and how we throw an exception if a method call is
+received for which we have no handler.
```

Comment:
Maybe better to code that using `response.error` as well. I think that's how one would do it on the iOS side because of the complexities of exception and memory handling.